From: srinivas pulukuru <srinivas.pulukuru@ti.com>
Date: Tue, 19 Apr 2011 06:21:07 +0000 (-0500)
Subject: OMAP4 DSS2 HDMI: Fix for HDMI disconnection during FIRST HPD processing
X-Git-Tag: android-2.6.35-2.3-omap4.13.1~28
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=2129803597344b3c8ca4a9789312243e7a690036

OMAP4 DSS2 HDMI: Fix for HDMI disconnection during FIRST HPD processing

After the FIRST HPD we usually see a disonnect interrupt coming before the
FIRST HPD interrupt is completely processed. IN this case FIRST HPD should
retyrn without configuring the HDMI and should not send HPD event to
userspace.
The IRQ's need to be reset before returning.

Change-Id: I5c0f7d86b8ab28505e81a571062f4a2234df7c26
Signed-off-by: srinivas pulukuru <srinivas.pulukuru@ti.com>
---

diff --git a/drivers/video/omap2/dss/hdmi.c b/drivers/video/omap2/dss/hdmi.c
index 068f387..32f880e 100644
--- a/drivers/video/omap2/dss/hdmi.c
+++ b/drivers/video/omap2/dss/hdmi.c
@@ -1485,15 +1485,21 @@ done:
 		 * the first time we enable HDMI via HPD.
 		 */
 		DSSINFO("Enabling display - HDMI_FIRST_HPD\n");
-		hdmi_reconfigure(dssdev);
 
 		if (!hdmi_connected) {
 			DSSINFO("irqstatus=0x%08x ignoring FIRST_HPD when "
 				"hdmi_connected = %d, hdmi_power = %d\n",
 				r, hdmi_connected, hdmi_power);
+			/* If HDMI is disconnected before FIRST HPD is processed
+			 * return without reconfiguring the HDMI and do not
+			 * send any hot plug event to the userspace in this case
+			 *  and reset irq's before returning.
+			 */
+			hdmi_set_irqs(0);
 			goto hpd_modify;
 		}
 
+		hdmi_reconfigure(dssdev);
 		set_hdmi_hot_plug_status(dssdev, true);
 		/* ignore return value for now */
 		DSSINFO("Enabling display Done- HDMI_FIRST_HPD\n\n");